servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpAutoDraining.java

```diff
@@ -0,0 +1,130 @@
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.TerminalSignalConsumer;
+import io.servicetalk.concurrent.internal.CancelImmediatelySubscriber;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.servicetalk.concurrent.api.Single.defer;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+public final class HttpAutoDraining implements StreamingHttpServiceFilterFactory {
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new DrainingStreamingHttpServiceFilter(service);
+    }
+
+    private static final class DrainingStreamingHttpServiceFilter extends StreamingHttpServiceFilter {
+
+        DrainingStreamingHttpServiceFilter(StreamingHttpService service) {
+            super(service);
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> handle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            return defer(() -> doHandle(ctx, request, responseFactory).shareContextOnSubscribe());
+        }
+
+        private Single<StreamingHttpResponse> doHandle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            final AtomicBoolean payloadSubscribed = new AtomicBoolean();
+            request.payloadBody().beforeOnSubscribe(ignored -> payloadSubscribed.set(true));
+            return delegate().handle(ctx, request, responseFactory).liftSync(subscriber ->
+                    new SingleSource.Subscriber<StreamingHttpResponse>() {
+
+                        // TODO: this could be a field with atomic field updater
+                        final AtomicBoolean cancellation = new AtomicBoolean();
+                        @Override
+                        public void onSubscribe(Cancellable cancellable) {
+                            subscriber.onSubscribe(() -> {
+                                try {
+                                    if (cancellation.compareAndSet(false, true)) {
+                                        maybeCancelMessageBody(payloadSubscribed, request.messageBody());
```

Comment:
Here you implemented "before cancel" behavior, while on payload body you used `afterFinally`. Do we really need "after" behavior or can we be happy with `BeforeFinallyHttpOperator`?
If "after" still makes sense, I see you added `AfterFinallyHttpOperator` in OTEL package. Should we just provide it as a utility in `http-utils`?